File: apps/windows/TeslaSync.App/feature-views/BatteryLevelChart.Model.cs
using TeslaSync.App.Core.Notifications;
using TeslaSync.App.Core.Units;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The mutually-exclusive render branch of the <c>BatteryLevelChart</c> surface — the native union of the
/// states the web component renders
/// (web/src/features/charging/components/charging-list/BatteryLevelChart.tsx). The web source is a pure
/// presentational component (it takes a single <c>data: StartLevelBucket[]</c> prop and performs no
/// fetching), so the branches are a direct function of the input <see cref="BatteryLevelChartModel"/> —
/// there is no fetch-driven error / stale / offline branch to reproduce here. The parent charging-list
/// experience owns the query lifecycle (loading / error / stale / offline are handled once for the whole
/// list before any chart is shown), exactly as the web page only renders this chart once the charging
/// sessions have resolved. Every branch maps onto a visible surface; none is ever hidden.
/// </summary>
public enum BatteryLevelChartState
{
    /// <summary>Initial fetch in flight (the parent is still loading the sessions) — title + skeleton chrome.</summary>
    Loading,

    /// <summary>Resolved with no charging activity to bucket — a friendly empty state, never a blank chart.</summary>
    Empty,

    /// <summary>At least one session falls into a starting-level bucket (web fall-through) — the bar strip.</summary>
    Ready,
}

/// <summary>
/// One starting-battery-level bucket — the native mirror of the web <c>StartLevelBucket</c> shape in
/// <c>web/src/features/charging/components/charging-list/helpers.ts</c>
/// (<c>{ range: string; count: number }</c>). <see cref="Range"/> is the 10-point SoC band label the web's
/// <c>computeStartLevelDist</c> emits (<c>"0-10%"</c> … <c>"90-100%"</c>); <see cref="Count"/> is the number
/// of charge sessions that started in that band. Pure data — no WinUI types.
/// </summary>
public sealed record BatteryLevelBucket(string Range, long Count);

/// <summary>
/// The render-time data model the <c>BatteryLevelChart</c> view binds to — the native analogue of the web
/// <c>BatteryLevelChartProps</c> (<c>data: StartLevelBucket[]</c>), plus the fetch flag the parent supplies
/// so the surface can render its own loading branch (the web parent gates the whole chart behind its query
/// state). The component is presentational: this model carries only the bucket series and the loading flag.
/// User-facing labels are resolved from the i18n facade by the projection, not passed in. Pure data — no
/// WinUI types — so the projection is unit-tested without a UI host.
/// </summary>
public sealed record BatteryLevelChartModel(bool Loading, IReadOnlyList<BatteryLevelBucket> Buckets)
{
    /// <summary>The initial model: the first fetch is in flight and no buckets have arrived yet.</summary>
    public static BatteryLevelChartModel Pending { get; } =
        new(true, Array.Empty<BatteryLevelBucket>());

    /// <summary>A resolved model with no buckets — the empty state.</summary>
    public static BatteryLevelChartModel Empty { get; } =
        new(false, Array.Empty<BatteryLevelBucket>());
}

/// <summary>
/// One projected, render-ready bar — the native analogue of a single recharts <c>&lt;Bar&gt;</c> datum.
/// <see cref="Range"/> is the SoC-band axis label shown beneath the bar (the web <c>XAxis dataKey="range"</c>);
/// <see cref="Count"/> is the raw session count; <see cref="CountText"/> is its grouped display form (the web
/// <c>fmtNumber</c> / <c>Intl.NumberFormat</c>); <see cref="HeightRatio"/> is the bar height as a fraction
/// (0..1) of the tallest bucket; and <see cref="AutomationName"/> is the spoken "{range}, {count} {sessions}"
/// the web tooltip conveys. Pure data.
/// </summary>
public sealed record BatteryLevelChartBar(
    string Range,
    long Count,
    string CountText,
    double HeightRatio,
    string AutomationName);

/// <summary>
/// The fully projected, render-ready view of the chart for one input model — the native analogue of what the
/// web <c>BatteryLevelChart</c> renders inside its <c>GlassPanel</c>. Holds the resolved title + hint (the two
/// <c>t(...)</c> calls from the web source), the empty + loading copy, the localized "sessions" word for the
/// bar Narrator names, the active <see cref="State"/>, the projected <see cref="Bars"/>, and the surface
/// <see cref="AutomationName"/>. Pure data so every branch is asserted headlessly.
/// </summary>
public sealed record BatteryLevelChartDisplay(
    BatteryLevelChartState State,
    string Title,
    string Hint,
    string EmptyMessage,
    string LoadingLabel,
    string SessionsWord,
    IReadOnlyList<BatteryLevelChartBar> Bars,
    string AutomationName);

/// <summary>
/// Pure projection from a <see cref="BatteryLevelChartModel"/> to its <see cref="BatteryLevelChartDisplay"/> —
/// the native port of web/src/features/charging/components/charging-list/BatteryLevelChart.tsx. The web
/// source is a bare titled bar chart of the start-of-charge SoC distribution; this projection reproduces its
/// two <c>t(...)</c> labels (<c>charging.charts.batteryLevelAtStart</c> / <c>charging.charts.batteryLevelHint</c>),
/// scales each bar to the tallest bucket, and formats the session counts through <see cref="NumberFormatting"/>
/// (the web <c>fmtNumber</c>). Because the web's <c>computeStartLevelDist</c> always emits ten buckets — every
/// one zero when there are no sessions — an all-zero (or absent) distribution carries no answer to "how low do
/// you typically go?", so it collapses to a friendly empty state per the surface state contract; the chart
/// renders as soon as any session lands in a bucket. Every label resolves through the i18n facade using the
/// same keys the web source feeds into <c>t(...)</c>. No WinUI types — unit-tested without a UI host.
/// </summary>
public static class BatteryLevelChartProjection
{
    private const string EmDash = "\u2014";

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web <c>data</c> prop + the parent's fetch flag).</param>
    /// <param name="localizer">The i18n facade every label resolves through.</param>
    public static BatteryLevelChartDisplay Project(BatteryLevelChartModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        string title = localizer.GetString("charging.charts.batteryLevelAtStart", "Battery Level at Charge Start");
        string hint = localizer.GetString(
            "charging.charts.batteryLevelHint",
            "How low do you typically go before charging?");
        string emptyMessage = localizer.GetString("chart.noData", "No data available");
        string loadingLabel = localizer.GetString("common.loading", "Loading");
        string sessionsWord = localizer.GetString("charging.charts.batteryLevelSessions", "Sessions");

        IReadOnlyList<BatteryLevelChartBar> bars = BuildBars(model.Buckets, sessionsWord);
        BatteryLevelChartState state = SelectState(model.Loading, model.Buckets);

        return new BatteryLevelChartDisplay(
            State: state,
            Title: title,
            Hint: hint,
            EmptyMessage: emptyMessage,
            LoadingLabel: loadingLabel,
            SessionsWord: sessionsWord,
            Bars: bars,
            AutomationName: BuildAutomationName(state, title, hint, emptyMessage, loadingLabel));
    }

    /// <summary>Branch precedence from the web source's data lifecycle: loading → empty → ready.</summary>
    private static BatteryLevelChartState SelectState(bool loading, IReadOnlyList<BatteryLevelBucket> buckets)
    {
        if (loading)
        {
            return BatteryLevelChartState.Loading;
        }

        // The web `computeStartLevelDist` always returns ten buckets (every count zero when there are no
        // sessions), so emptiness is a function of the VALUES, not the bucket count: a distribution with no
        // sessions in any band answers nothing and collapses to the empty state, while a single session in
        // any band promotes the surface to the charted Ready state.
        foreach (var bucket in buckets)
        {
            if (bucket.Count > 0)
            {
                return BatteryLevelChartState.Ready;
            }
        }

        return BatteryLevelChartState.Empty;
    }

    private static IReadOnlyList<BatteryLevelChartBar> BuildBars(
        IReadOnlyList<BatteryLevelBucket> buckets,
        string sessionsWord)
    {
        if (buckets.Count == 0)
        {
            return Array.Empty<BatteryLevelChartBar>();
        }

        long max = 0;
        foreach (var bucket in buckets)
        {
            if (bucket.Count > max)
            {
                max = bucket.Count;
            }
        }

        var bars = new List<BatteryLevelChartBar>(buckets.Count);
        foreach (var bucket in buckets)
        {
            string range = string.IsNullOrWhiteSpace(bucket.Range) ? EmDash : bucket.Range;
            long count = bucket.Count < 0 ? 0 : bucket.Count;
            string countText = NumberFormatting.Format(count, null, 0);
            double ratio = max > 0 ? Math.Clamp(count / (double)max, 0.0, 1.0) : 0.0;

            bars.Add(new BatteryLevelChartBar(
                Range: range,
                Count: count,
                CountText: countText,
                HeightRatio: ratio,
                AutomationName: $"{range}, {countText} {sessionsWord}"));
        }

        return bars;
    }

    private static string BuildAutomationName(
        BatteryLevelChartState state,
        string title,
        string hint,
        string emptyMessage,
        string loadingLabel) => state switch
        {
            BatteryLevelChartState.Loading => $"{title}. {loadingLabel}",
            BatteryLevelChartState.Empty => $"{title}. {emptyMessage}",
            _ => $"{title}. {hint}",
        };
}

/// <summary>
/// PII-safe diagnostics for the <c>BatteryLevelChart</c> surface (P1/S11 diagnostics contract). Records only
/// the operational <c>view.opened</c> event with the surface slug — never a bucket label or session count —
/// so a diagnostics line can never leak a user's charging behaviour. Thread-safe.
/// </summary>
public sealed class BatteryLevelChartDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public BatteryLevelChartDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=BatteryLevelChart</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={BatteryLevelChartRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>BatteryLevelChart</c> feature surface — the native mirror of the web
/// component at <c>web/src/features/charging/components/charging-list/BatteryLevelChart.tsx</c>.
/// </summary>
public static class BatteryLevelChartRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "BatteryLevelChart";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/BatteryLevelChart/BatteryLevelChartModel.kt
// Pure, framework-free model + projection for the Battery-Level-at-Charge-Start chart feature view — the
// native analogue of everything the web surface derives before returning JSX
// (web/src/features/charging/components/charging-list/BatteryLevelChart.tsx, fed by the sibling
// helpers.ts `computeStartLevelDist`). No Compose, no Android, no HTTP: every declaration here is
// unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the composable a thin render
// layer.
//
// The web component is purely presentational — its parent computes `computeStartLevelDist(sessions)` and
// passes the resulting `StartLevelBucket[]` (ten fixed 0-10% … 90-100% start-of-charge SoC buckets, each
// with a session count) as the `data` prop; the component then draws one amber bar per bucket. This file
// owns both halves of that contract: [BatteryLevelChartProjection.distribution] mirrors the web
// `computeStartLevelDist` (the client-side bucketing the web parent runs), and
// [BatteryLevelChartProjection.projectBuckets] maps those buckets into the render-ready chart inputs the
// composable consumes (the X-axis range labels, the bar values, and the accessible fallback-table rows).
// Bucket order is the natural ascending 0→90% order the web array builder produces, so the native bars and
// the accessible table read left-to-right in the same order.
//
// SI on the wire, display at the boundary: `startSocPct` is a battery state-of-charge percentage (0-100,
// dimensionless) exactly as the API serves it — there is no unit conversion to do, only bucketing, which
// happens here.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/BatteryLevelChart — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting types.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.batterylevelchart

import io.teslasync.android.data.UiState
import io.teslasync.shared.core.diagnostics.Logger
import java.util.Locale
import kotlin.math.floor

/** Number of start-of-charge SoC buckets — the web `Array.from({ length: 10 })`. */
internal const val BUCKET_COUNT: Int = 10

/** Width of each SoC bucket in percentage points — the web `i * 10` / `+ 10` step. */
internal const val BUCKET_WIDTH_PCT: Int = 10

/** Highest bucket index — the web `Math.min(idx, 9)` clamp upper bound. */
internal const val MAX_BUCKET_INDEX: Int = BUCKET_COUNT - 1

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object BatteryLevelChartRegistration {
    /** Stable surface id. */
    const val ID: String = "battery-level-chart"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "BatteryLevelChart"
}

/**
 * The subset of a charging session this surface reads — the native mirror of the single web
 * `ChargingSession` field the distribution touches. [startSocPct] is the battery state of charge at the
 * moment charging began, as a 0-100 percentage (web `start_soc_pct`).
 */
data class ChargingSessionStart(
    val startSocPct: Double,
)

/**
 * One start-of-charge SoC bucket — the native mirror of the web `StartLevelBucket`
 * (`{ range: string; count: number }`). [range] is the human-readable band label (e.g. `"0-10%"`) and
 * [count] is how many sessions began in that band (a non-negative integer, so `Long`).
 */
data class StartLevelBucket(
    val range: String,
    val count: Long,
)

/**
 * The already-localized chart microcopy the composable reads from the i18n catalog (P1/S10) — the two
 * `charging.charts.*` keys the web component resolves via `t(...)` plus the accessible table column
 * headers and the bar series name (web `<Bar name="Sessions" />`). The lifecycle-chrome strings
 * (empty / error / retry / offline / freshness) are resolved inline at the Compose boundary, not here, so
 * this holder stays a thin content carrier.
 */
data class BatteryLevelChartStrings(
    val title: String,
    val subtitle: String,
    val rangeColumn: String,
    val sessionsColumn: String,
    val seriesLabel: String,
)

/**
 * The fully projected, render-ready chart inputs — the native analogue of the props the web `<BarChart>` +
 * its parent read from the `StartLevelBucket[]`. Pure data (no Compose types) so the projection is
 * unit-tested without a UI host: the composable wraps [values] into a single `ChartSeries`, feeds
 * [xLabels] to the bottom axis, renders [tableRows] as the accessible fallback table, and shows the
 * friendly empty state when [isEmpty] (no session fell into any bucket).
 */
data class BatteryLevelChartProjectionResult(
    val xLabels: List<String>,
    val values: List<Double>,
    val tableRows: List<List<String>>,
    val isEmpty: Boolean,
)

/**
 * The pure projection the composable renders — the native mirror of the web `computeStartLevelDist`
 * helper and the component's read of its result. Stateless and side-effect-free so it is fully covered by
 * the off-device unit gate.
 */
object BatteryLevelChartProjection {
    /**
     * The bucket index a [startSocPct] falls into — the native mirror of the web
     * `Math.min(Math.floor(start_soc_pct / 10), 9)`. The result is additionally floored at 0 so a
     * malformed negative SoC clamps into the first band instead of throwing (the web indexes an array
     * directly); a 100% start lands in the top `90-100%` band.
     */
    fun bucketIndex(startSocPct: Double): Int = floor(startSocPct / BUCKET_WIDTH_PCT).toInt().coerceIn(0, MAX_BUCKET_INDEX)

    /**
     * Buckets [sessions] into the ten fixed start-of-charge SoC bands — the native mirror of the web
     * `computeStartLevelDist`. Always returns [BUCKET_COUNT] ascending buckets (`0-10%` … `90-100%`),
     * each carrying the number of sessions that began in that band, so the chart's x-axis is stable
     * regardless of how sparse the data is.
     */
    fun distribution(sessions: List<ChargingSessionStart>): List<StartLevelBucket> {
        val counts = LongArray(BUCKET_COUNT)
        for (session in sessions) {
            counts[bucketIndex(session.startSocPct)] += 1L
        }
        return (0 until BUCKET_COUNT).map { index ->
            val low = index * BUCKET_WIDTH_PCT
            StartLevelBucket(range = "$low-${low + BUCKET_WIDTH_PCT}%", count = counts[index])
        }
    }

    /**
     * Projects pre-computed [buckets] into render-ready chart inputs, preserving order. Each bucket
     * contributes one X-axis range label, one bar value (its count), and one accessible-table row
     * (`[range, formatCount(count)]`). [isEmpty] is true only when no bucket holds a session — ten
     * all-zero bars would read as a blank panel, so the surface shows the friendly empty state instead.
     * Injecting [formatCount] keeps this locale-deterministic under test.
     */
    fun projectBuckets(
        buckets: List<StartLevelBucket>,
        formatCount: (count: Long) -> String,
    ): BatteryLevelChartProjectionResult =
        BatteryLevelChartProjectionResult(
            xLabels = buckets.map { it.range },
            // `+ 0.0` widens each Long band count to the chart series' Double value type.
            values = buckets.map { it.count + 0.0 },
            tableRows = buckets.map { listOf(it.range, formatCount(it.count)) },
            isEmpty = buckets.all { it.count == 0L },
        )

    /**
     * Convenience that buckets [sessions] (via [distribution]) and projects the result in one step — the
     * full web data flow (`computeStartLevelDist` → render) the stateful entry binds to the charging feed.
     */
    fun project(
        sessions: List<ChargingSessionStart>,
        formatCount: (count: Long) -> String,
    ): BatteryLevelChartProjectionResult = projectBuckets(distribution(sessions), formatCount)

    /** Locale-grouped integer formatting — the native analogue of the web `fmtInt` (e.g. `1,204`). */
    fun formatCount(
        count: Long,
        locale: Locale = Locale.getDefault(),
    ): String = String.format(locale, "%,d", count)
}

/**
 * Maps a charging-feed [UiState] of raw [ChargingSessionStart]s onto the buckets the surface renders,
 * running [BatteryLevelChartProjection.distribution] over the payload while preserving every lifecycle
 * field (phase, freshness stamp, stale/refreshing flags, error classification). This is how the stateful
 * sessions entry reproduces the web parent's `computeStartLevelDist(sessions)` without losing the
 * cache-then-network contract. A `null` payload (first load / hard error) stays `null`.
 */
fun distributionState(state: UiState<List<ChargingSessionStart>>): UiState<List<StartLevelBucket>> =
    UiState(
        phase = state.phase,
        data = state.data?.let { BatteryLevelChartProjection.distribution(it) },
        fetchedAt = state.fetchedAt,
        stale = state.stale,
        refreshing = state.refreshing,
        errorKind = state.errorKind,
        httpStatus = state.httpStatus,
    )

/**
 * Emits the one PII-safe `view.opened` diagnostic with the surface [BatteryLevelChartRegistration.SLUG]
 * (P1/S11). Kept free of Compose so it is unit-tested with a recording [Logger]; the composable calls it
 * from its first-composition effect.
 */
fun recordBatteryLevelChartOpened(logger: Logger) {
    logger.info("view.opened", mapOf("surface" to BatteryLevelChartRegistration.SLUG))
}
